### PR TITLE
test(hooks): increase coverage

### LINKED ISSUE #3381: Add unit test for hooks
## Why

Hooks are our connection to the backend and the is some logic in them that needs to be tested to ensure that the edge cases work propertly.

## Tools to implement them

Since vitest is already part of the project, it will be the perfect tool to develop the unit tests.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by paaragon**
## Changes

Closes #3381 

It adds tests to cover everything under hooks/

## Implementation details

Test are done with vitest, which is the current test library in the project.

### CODE REVIEW

**Comment by paaragon** [comment]
image

File: tests/hooks/event-conversion-utils.spec.ts [hunk 1]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
+    } catch (e) {
+      expect((e as any).message).toBe(
+        "Unreachable: should have all the cases covered."
+      );
+    }
+  });
+});
+
+describe("getEventsQuery", () => {
+  test("check query arguments", () => {
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:821]:
```suggestion
  test("uses expected arguments for the Airtable query", () => {
```

File: tests/hooks/event-conversion-utils.spec.ts [hunk 2]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
+          case "End date":
+            return "2020-03-02";
+          case "Website":
+            return "https://qiskit.org/events";
+        }
+      },
+    },
+  ];
+
+  test("fetch in date range events", async () => {
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:896]:
```suggestion
  test("fetches events in date range", async () => {
```

File: tests/hooks/event-conversion-utils.spec.ts [hunk 3]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
+        startDate: "2020-03-01",
+        startDateAndTime: null,
+        title: "Fake conference 2",
+        to: "https://qiskit.org/events",
+        types: ["Hackathon"],
+      },
+    ]);
+  });
+
+  test("fetch in date range events sorted descending", async () => {
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:951]:
```suggestion
  test("fetches events in date range and sorts them in descending order", async () => {
```

File: tests/hooks/event-conversion-utils.spec.ts [hunk 4]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
+        to: "http://overlooked-sleeping.org",
+        abstract: "",
+        date: "July 29, 2022",
+        image: "/images/events/downloaded/undefined.jpg",
+        startDate: "2022-07-29",
+      },
+    ]);
+  });
+
+  test("If no recordFields should get from getAllFieldNames", async () => {
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:1165]:
```suggestion
  test("gets record fields from `getAllFieldNames` if there are none yet", async () => {
```

File: tests/hooks/event-conversion-utils.spec.ts [hunk 5]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
+
+    eventsAirtableRecords.getEventsQuery = getEventsQueryMock;
+    eventsAirtableRecords.getAllFieldNames = getAllFieldNamesMock;
+
+    await eventsAirtableRecords.fetchSeminarSeriesEvents(10);
+
+    expect(eventsAirtableRecords.recordFields).toStrictEqual(mockRecordFields);
+  });
+
+  test("test fallback cases", async () => {
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:1189]: I think that we should explain what specific fallback cases we're testing here.

File: tests/hooks/event-conversion-utils.spec.ts [hunk 6]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
+      speaker: "",
+      startDate: "",
+      title: "",
+      to: "",
+    });
+  });
+});
+
+describe("formatTime", () => {
+  test("format a valid date", () => {
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:1216]:
```suggestion
  test("formats date", () => {
```

File: tests/hooks/event-conversion-utils.spec.ts [hunk 7]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
+    const eventsAirtableRecords = new FakeEventsAirtableRecords(
+      "testApiKey",
+      "testView"
+    );
+
+    eventsAirtableRecords.recordFields = {};
+
+    const date = eventsAirtableRecords.formatTime("2020-10-10 00:00:00 UTC");
+
+    expect(date.replace("\u202F", " ")).toBe("12:00 AM UTC");
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:1226]: Is there maybe a way of writing this assertion without modifying the value returned from the tested method (`date`)? Otherwise it feels _hacky_.

File: tests/hooks/mock/seminarSeriesMock.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
+export const mock: any[] = [
+  {
+    Name: "delectus eaque consectetur",
+    "Event Start Date": "2022-07-29",
+    "Type of Engagement": ["Seminar Series"],
+    "Event Website": "http://overlooked-sleeping.org",
```
Comment (eddybrando) [inline tests/hooks/mock/seminarSeriesMock.ts:6]: It might be safer to replace all mock URLs with `https://example.com/` or `https://example.org/` (a path can be appended), since it's a safe reserved URL for testing.

Relevant documents:

- https://www.rfc-editor.org/rfc/rfc2606.html
- https://www.rfc-editor.org/rfc/rfc6761.html

File: tests/hooks/utils/conversion-utils.spec.ts [hunk 9]
```diff
@@ -1,35 +1,93 @@
...
 import { writeJSONToFile } from "../../../hooks/utils/conversion-utils";
 
 describe("writeJSONToFile", () => {
-  test("write a dummy json to a file", async () => {
+  beforeEach(() => {
+    vi.clearAllMocks();
+    vi.mock("fs");
+  });
+
+  test("write a file in an unexistent folder", async () => {
```
Comment (eddybrando) [inline tests/hooks/utils/conversion-utils.spec.ts:12]:
```suggestion
  test("writes a file in an nonexistent directory", async () => {
```

File: tests/hooks/utils/conversion-utils.spec.ts [hunk 10]
```diff
@@ -1,35 +1,93 @@
...
+
     const mock = [
       {
-        name: "Main",
-        description: "The main Qiskit packages maintained by IBM Quantum",
-      },
-      {
-        name: "Extensions",
-        description: "IBM Quantum supported Qiskit extensions",
+        name: "Abby Mitchell",
```
Comment (eddybrando) [inline tests/hooks/utils/conversion-utils.spec.ts:21]: It's a good idea to avoid real names in mock data.

File: tests/hooks/utils/conversion-utils.spec.ts [hunk 11]
```diff
@@ -1,35 +1,93 @@
...
+    expect(fs.promises.stat).toBeCalledWith(folder);
+    expect(fs.promises.mkdir).toBeCalledWith(folder);
+    expect(fs.promises.writeFile).toBeCalledWith(
+      filePath,
+      JSON.stringify(mock, null, 2)
+    );
+  });
 
-    const mockPath = path.join(__dirname, "mock.json");
+  test("write a file in an unexistent folder", async () => {
```
Comment (eddybrando) [inline tests/hooks/utils/conversion-utils.spec.ts:54]:
```suggestion
  test("writes a file in an existent directory", async () => {
```

**Review by eddybrando — CHANGES_REQUESTED** [review]
Awesome work here @paaragon! You really got to cover all possible relevant cases 💪

Please take a look at my suggested changes.

File: tests/hooks/airtable-conversion-utils.spec.ts [hunk 12]
```diff
@@ -62,4 +75,207 @@ describe("storeImage", () => {
...
+      const result = await airtableRecords.getAllFieldNames(RECORD_FIELDS_IDS);
+      // notice that in this object, there is no name because is th one that
+      // throws the exception
+      expect(result).toStrictEqual({
+        city: "City",
+        country: "City",
+        region: "City",
+        image: "City",
+        slackUsername: "City",
+        slackId: "City",
```
[the remark this answers is not in the document]
Comment (paaragon) [reply tests/hooks/airtable-conversion-utils.spec.ts]: Fixed! thanks @eddybrando

File: out.txt
[the remark this answers is not in the document]
Comment (paaragon) [reply out.txt:1]: Removed!

File: tests/hooks/event-conversion-utils.spec.ts (hunk 7 above)
Comment (paaragon) [reply tests/hooks/event-conversion-utils.spec.ts:1226]: Is an issue about platforms (related to encoding). On my local, date has this character `\u202F` but in CI it doesn't

File: tests/hooks/mock/seminarSeriesMock.ts (hunk 8 above)
Comment (paaragon) [reply tests/hooks/mock/seminarSeriesMock.ts:6]: Thanks for the info!

File: tests/hooks/utils/conversion-utils.spec.ts (hunk 10 above)
Comment (paaragon) [reply tests/hooks/utils/conversion-utils.spec.ts:21]: Done

File: tests/hooks/airtable-conversion-utils.spec.ts [hunk 13]
```diff
@@ -62,4 +75,207 @@ describe("storeImage", () => {
...
+    });
+
+    test("If a field causes a handled error, it shouln'd appear in the final object", async () => {
+      const airtableEachPageMockFn = vi
+        .fn()
+        .mockImplementation(
+          (cb: (records: any, nextPage: any) => void): Promise<void> => {
+            // eslint-disable-next-line n/no-callback-literal
+            cb(
+              // two elements to check the  for loop in getFieldName
```
[the remark this answers is not in the document]
Comment (eddybrando) [reply tests/hooks/airtable-conversion-utils.spec.ts:208]: I believe this wasn't changed yet 😬

File: tests/hooks/airtable-conversion-utils.spec.ts [hunk 14]
```diff
@@ -62,4 +75,228 @@ describe("storeImage", () => {
...
+      ];
+      const airtableEachPageMockFn = vi
+        .fn()
+        .mockImplementation(
+          (cb: (records: any, nextPage: any) => void): Promise<void> => {
+            // this should pass 0 records for the first field (which is name)
+            const records = callIndex === 0 ? [] : [fields[callIndex]];
+            callIndex++;
+            cb(
+              // two elements to check the  for loop in getFieldName
```
Comment (eddybrando) [inline tests/hooks/airtable-conversion-utils.spec.ts:266]:
```suggestion
              // two elements to check the for loop in `getFieldName`
```

File: tests/hooks/event-conversion-utils.spec.ts (hunk 5 above)
Comment (paaragon) [reply tests/hooks/event-conversion-utils.spec.ts]: I changed the description

File: tests/hooks/event-conversion-utils.spec.ts [hunk 15]
```diff
@@ -707,9 +803,426 @@ describe("formatDates", () => {
...
-      "January 1-2, 2020"
+  test("test practial unreachable case", () => {
+    try {
+      eventsAirtableRecords.formatDates(
+        new Date("2020-01-01T00:00:00.000Z"),
+        new Date("2020-01-01T00:10:00.000Z")
+      );
+    } catch (e) {
+      expect((e as any).message).toBe(
+        "Unreachable: should have all the cases covered."
```
[the remark this answers is not in the document]
Comment (eddybrando) [reply tests/hooks/event-conversion-utils.spec.ts:814]: I think that this one here hasn't been addressed yet.

File: tests/hooks/event-conversion-utils.spec.ts [hunk 16]
```diff
@@ -707,9 +803,427 @@ describe("formatDates", () => {
...
+
+    eventsAirtableRecords.getEventsQuery = getEventsQueryMock;
+    eventsAirtableRecords.getAllFieldNames = getAllFieldNamesMock;
+
+    await eventsAirtableRecords.fetchSeminarSeriesEvents(10);
+
+    expect(eventsAirtableRecords.recordFields).toStrictEqual(mockRecordFields);
+  });
+
+  test("Each field has a default value if `record.get(field)` returns falsy", async () => {
```
Comment (eddybrando) [inline tests/hooks/event-conversion-utils.spec.ts:1189]:
```suggestion
  test("sets default values for each field if `get(field)` is falsy", async () => {
```

File: tests/hooks/event-conversion-utils.spec.ts (hunk 7 above)
Comment (eddybrando) [reply tests/hooks/event-conversion-utils.spec.ts]: It makes sense, thank you for the explanation.

Maybe we could go for one of these approaches:

1. Add a comment to describe why we need to do this `replace` to avoid confusion in the future.
2. Change the assertion to [`toContain`](https://jestjs.io/docs/expect#tocontainitem) instead of `toBe`. (If this works, this would be my personal preference)

File: tests/hooks/mock/seminarSeriesMock.ts [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    "Event City/State": "YouTube",
+    "Event Abstract": "Qui facere assumenda ducimus.",
+    "Add to Event Site": true,
+    "Event Website": "http://example.com",
+    "Image / Icon": [
+      {
+        id: "att8eF6mVuKzbZxk0",
+        width: 604,
+        height: 684,
+        url: "http://dimwitted-claw.org",
```
Comment (eddybrando) [inline tests/hooks/mock/seminarSeriesMock.ts:55]: It might be safer to replace all mock URLs with `https://example.com/` or `https://example.org/` (a path can be appended), since it's a safe reserved URL for testing.

Relevant documents:

- https://www.rfc-editor.org/rfc/rfc2606.html
- https://www.rfc-editor.org/rfc/rfc6761.html

File: tests/hooks/mock/seminarSeriesMock.ts [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        filename: "intelligent_deposit.sass",
+        size: 400718,
+        type: "image/png",
+        thumbnails: {
+          small: {
+            url: "http://amused-microwave.org",
```
Comment (eddybrando) [inline tests/hooks/mock/seminarSeriesMock.ts:61]: It might be safer to replace all mock URLs with `https://example.com/` or `https://example.org/` (a path can be appended), since it's a safe reserved URL for testing.

Relevant documents:

- https://www.rfc-editor.org/rfc/rfc2606.html
- https://www.rfc-editor.org/rfc/rfc6761.html

File: tests/hooks/utils/conversion-utils.spec.ts [hunk 19]
```diff
@@ -1,35 +1,93 @@
...
+        city: "New York",
+        country: ["United States"],
+        slackId: "U017XULSMUY",
+        slackUsername: "Maria.Dickens34",
       },
       {
-        name: "Community",
-        description:
-          "Software packages supported by the Qiskit community, not maintained by IBM Quantum",
+        name: "Abeer Vaishnav",
```
Comment (eddybrando) [inline tests/hooks/utils/conversion-utils.spec.ts:30]: It's a good idea to avoid real names in mock data.

Some _safe_ names to use are **John Doe** and **Jane Doe** for instance: https://en.wikipedia.org/wiki/John_Doe

**Review by eddybrando — CHANGES_REQUESTED** [review]
Thank you very much for the changes @paaragon!

I've made another review and added a couple of comments.

File: tests/hooks/event-conversion-utils.spec.ts (hunk 7 above)
Comment (paaragon) [reply tests/hooks/event-conversion-utils.spec.ts]: Great! thanks for those tips

File: tests/hooks/utils/conversion-utils.spec.ts (hunk 19 above)
Comment (paaragon) [reply tests/hooks/utils/conversion-utils.spec.ts:30]: I think now all names are replaced by Mr and Mrs Doe

File: tests/hooks/mock/seminarSeriesMock.ts (hunk 17 above)
Comment (paaragon) [reply tests/hooks/mock/seminarSeriesMock.ts]: Done!

File: tests/hooks/mock/seminarSeriesMock.ts (hunk 18 above)
Comment (paaragon) [reply tests/hooks/mock/seminarSeriesMock.ts]: Done

**Review by eddybrando — APPROVED** [review]
Great work here @paaragon! 💪
